### ISSUE TITLE
Vinyl stalls on intensive random insertion

### ISSUE THREAD

**Issue by funny-falcon** [tarantool/tarantool#3603]
https://gist.github.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4

```
$ go run test_tarantool.go -conc 2000 -rw 0
...
96207.21 rps
114223.01 rps
83925.10 rps
58003.15 rps
0.00 rps
131.02 rps
0.00 rps
0.00 rps
0.00 rps
0.00 rps
0.00 rps
0.00 rps
0.00 rps
2018/08/03 12:39:39 client timeout for request 9777472 (0x4003)
2018/08/03 12:39:39 client timeout for request 9777616 (0x4003)
exit status 1
```

**Comment by locker** [comment]
Looks like this happens, because we don't have tx throttling: #1862

**Comment by funny-falcon** [comment]
1. latest commit 2.0 branch
2. 4 cores HT (8 virtual cores).
16GB RAM (although some consumed by browser/editor/etc, but at least 10GB available) .
Consumer grade, but quality SATA SSD (Dell's native option for workstations).
3. Go 1.10.1 . I didn't change GOMAXPROCS.
4. Ok, I will run it again.

Are this questions because issue is not reproducible in your environment?

**Comment by funny-falcon** [comment]
Oh, sorry, that was Tarantool 1.10.1-254-gac590da85 from branch 1.10.

I've attached [z_tarantool.log](https://gist.github.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/86787d18eda92a90e049fe7d8022dfdd01c374be/z_tarantool.log) and [z_test_tarantool.out](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/86787d18eda92a90e049fe7d8022dfdd01c374be/z_test_tarantool.out) to [gist](https://gist.github.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4)

**Comment by funny-falcon** [comment]
SSD disk capable to 40k random iops as fio reported.
I runned on encrypted volume and encrypted folder, therefore, performance was lower, and directio doesn't work.

I've rerunned on non-encrypted partition.
It takes longer time to reproduce, but still after 6 minutes it failed (ie 10 second timeout exceeded).

**Comment by locker** [comment]
@funny-falcon thanks for the update.

> Are this questions because issue is not reproducible in your environment?

I couldn't reproduce the issue on my laptop (which is rather sluggish I must admit) or on a test server (Xeon 24 logical CPUs, 100 GB of RAM). However, both machines have rotational disks.

_Well, I could get the client timeout, but it took several hours and it happened, because tarantool ran out of disk space, but apparently this isn't what happened in your case (it would be obvious from the log)._

> I've attached z_tarantool.log and z_test_tarantool.out to gist

The issue has nothing to do with vinyl memory quota (and so throttling wouldn't help), otherwise we would see a lot of messages like the one below (there's none in the log you attached):
```
vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 0.535 sec
```

What I can see is tons of messages similar to this one:
```
txn.c:295 W> too long WAL write: 1 rows at LSN 8093436: 0.852 sec
```

At the first glance, looks like vinyl consumed all available disk bandwidth while doing dump/compaction, which caused WAL writer starvation. Ideally, WAL files (_xlog_; see `box.cfg.wal_dir`) and data files (_snap, run_; see `box.cfg.memtx_dir`, `box.cfg.vinyl_dir`) should be stored on different physical devices. If it is not an option, you can try to tweak `box.cfg.snap_io_rate_limit` (it limits the bandwidth of both memtx snapshot writer and vinyl dump/compaction tasks) and `box.cfg.vinyl_write_threads` (set it to minimum 2). This can result in vinyl memory quota errors I mentioned above though.

However, what is curious, I saw lots of such messages when I ran the test on my machine and I did experience RPS drops, but they were minor, not enough to trigger timeout. So the issue may also be caused by your SSD drive. May be, you don't have [TRIM](https://en.wikipedia.org/wiki/Trim_(computing)) enabled or the IO scheduler is misconfigured? (Just guessing, I'm not an expert in this kind of thing).

Another interesting thing is that I saw `too long WAL write` warnings (and hence RPS drops) even after I increased `box.cfg.vinyl_memory` up to 20 GB and reset `box.cfg.checkpoint_interval` to the default value (1 hour; you have 10 seconds) or even completely switched to memtx. That is, the issue manifests itself even if WAL thread is the only disk writer. I'm going to investigate this. Looks like it may be unrelated to vinyl, after all. Could you rerun your test on memtx to check that? (Don't forget to reset `box.cfg.checkpoint_interval` to make sure snapshot writer doesn't eat up all disk bandwidth).

**Comment by locker** [comment]
@funny-falcon Here's what happened on my hardware. Since my server has a lot of memory, the OS sets a rather lenient dirty memory limit (it is configured as some percentage of total memory). So tarantool manages to accumulate a lot of dirty pages before they are finally flushed to disk. As a result, flushing takes quite some time, consuming all disk bandwidth and stalling tarantool. I lowered the dirty memory limit down to 10 MB and stalls disappeared. Could you try it out and report back if it helps in your case?
```
# sysctl vm.dirty_bytes=$[10<<20]
```

**Comment by funny-falcon** [comment]
> The issue has nothing to do with vinyl memory quota (and so throttling wouldn't help), otherwise we would see a lot of messages like the one below (there's none in the log you attached):
> `vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 0.535 sec`

Damn :-( looks like I didn't copy whole log :-( most probably, clipboard buffer is too small for them. And Gist doesn't allow to attach buffer in web interface.

In fact there were a lot of such records.

I will rerun that again. And will try to attach files to gist using git.
Excuse me for that.

**Comment by funny-falcon** [comment]
btw, it is a really bad idea to limit down `vm.dirty_(bytes|ratio)`.
It is much better to set `vm.dirty_background_(bytes|ratio)`. And also reduce `vm.dirty_expire_centisecs` and `vm.dirty_writeback_centisecs`.

**Comment by funny-falcon** [comment]
Ok, I've runned again.
- [xa_tarantool.log](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/c23f04dde67ade7a1828f15b8de974d98f79da4b/xa_tarantool.log) , [xa_test_tarantool.out](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/c23f04dde67ade7a1828f15b8de974d98f79da4b/xa_test_tarantool.out) - from default on non-encrypted partition
- [xb_tarantool.log](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/c23f04dde67ade7a1828f15b8de974d98f79da4b/xb_tarantool.log), [xb_test_tarantool.out](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/c23f04dde67ade7a1828f15b8de974d98f79da4b/xb_test_tarantool.out) - with `vm.dirty_background_ratio=1`, `vm.dirty_expire_centisecs=100` and `vm.dirty_writeback_centisecs=10`.
- [xc_tarantool.log](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/c23f04dde67ade7a1828f15b8de974d98f79da4b/xc_tarantool.log), [xc_test_tarantool.out](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/c23f04dde67ade7a1828f15b8de974d98f79da4b/xc_test_tarantool.out) - same as previous, but also set `vm.dirty_bytes=$((10<<20))` as you recommended. It runs for 13 minutes with longest stall for 7 seconds. Much better, but 7 seconds is still too much for production. And setting `vm.dirty_bytes` to such low value is still bad decision, imho.

At this moment I recognize, that snapshoting is a trigger. You can see it also by comparing times of stallness in test_tarantool.out and saving snapshots in tarantool.log

- [xd_tarantool.log](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/c23f04dde67ade7a1828f15b8de974d98f79da4b/xd_tarantool.log), [xd_test_tarantool.out](https://gist.githubusercontent.com/funny-falcon/5d1ec0e49ec881c68722335ffb78cca4/raw/c23f04dde67ade7a1828f15b8de974d98f79da4b/xd_test_tarantool.out) - same as before, but also set `snap_io_rate_limit=10` in config.lua (documentation says, it is value in MB/sec). As you can see, things became worse. I previously tested `snap_io_rate_limit=30` and it certainly were worse than without setting this value. And `snap_io_rate_limit=10` is worse than `snap_io_rate_limit=30`. And I also repeated it with default sysctl parameters. Result is same.

Note: 10 seconds is just to demonstrate awfulness of issue. In practice, even 2 seconds is already issue.

checkpoint_interval is so low because otherwise log files quickly eats whole space. Any way, snapshot should not trigger such huge delay even being on the same disk. It doesn't matter, happens it six times a minute or once a hour - issue is the issue.
At least, there should be workable configuration options **to throttle snapshot and to not throttle vynil at the same time**.

You can see that i had commented `vinyl_write_threads` in config.lua . So all logs are with default value (it should be equal to 2 according to documentation).

**Comment by locker** [comment]
@funny-falcon Thanks a lot for the update, you've been really helpful. Looks like the problem is caused (at least partially) by absence of throttling in vinyl. We will escalate the priority of the corresponding issue (#1862).

For the record, what I see from the logs you provided:
- _Default setup._ Both vinyl memory quota errors and WAL errors are present in the log. Last distinct errors before the timeout:
```
2018-08-09 19:41:01.324 [11818] main/1692/main txn.c:295 W> too long WAL write: 1 rows at LSN 35915253: 1.715 sec
2018-08-09 19:41:02.486 [11818] main/303/main vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 0.934 sec
2018-08-09 19:41:34.366 [11818] main/1267/main txn.c:295 W> too long WAL write: 1 rows at LSN 38493255: 1.838 sec
2018-08-09 19:41:46.763 [11818] main/616/main vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 10.542 sec
```
- _vm.dirty_background_ratio=1, vm.dirty_expire_centisecs=100, and vm.dirty_writeback_centisecs=10._
Again, both vinyl memory quota errors and WAL errors are present in the log, but WAL timeouts look less severe. Last distinct errors before the timeout:
```
2018-08-09 19:49:26.045 [13442] main/1894/main txn.c:295 W> too long WAL write: 1 rows at LSN 11639470: 0.543 sec
2018-08-09 19:49:17.783 [13442] main/1278/main vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 2.652 sec
2018-08-09 19:49:26.045 [13442] main/1894/main txn.c:295 W> too long WAL write: 1 rows at LSN 11639470: 0.543 sec
2018-08-09 19:50:03.522 [13442] main/2041/main vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 15.917 sec
```
- _Same as before + vm.dirty_bytes=$((10<<20))._ There are many fewer WAL errors; actually, there's just one distinct error:
```
2018-08-09 20:09:23.756 [15858] main/674/main txn.c:295 W> too long WAL write: 1 rows at LSN 34206947: 0.707 sec
```
There are as many vinyl memory quota errors as before, but they look less severe:
```
2018-08-09 20:14:22.177 [15858] main/2045/main vy_quota.h:170 W> waited for 554 bytes of vinyl memory quota for too long: 3.206 sec
2018-08-09 20:14:57.243 [15858] main/157/main vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 4.955 sec
2018-08-09 20:15:26.645 [15858] main/678/main vy_quota.h:170 W> waited for 554 bytes of vinyl memory quota for too long: 2.848 sec
2018-08-09 20:16:01.857 [15858] main/1731/main vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 6.053 sec
```
- _Same as before plus snap_io_rate_limit=10._ There are no WAL errors, but vinyl memory quota timeouts are much higher:
```
2018-08-09 20:26:45.334 [18349] main/1827/main vy_quota.h:170 W> waited for 555 bytes of vinyl memory quota for too long: 31.584 sec
```

**Comment by locker** [comment]
@funny-falcon Could you please also re-run the test for a memtx space? I ask, because there seems to be two issues here: absence of throttling in vinyl and absence of throttling in WAL (or may be sub-optimal syncing?). I'd appreciate if you could answer the questions below:
- Do you experience the same stalls on memtx?
- If yes, can you mitigate them by tweaking sysctl knobs (vm.dirty_bytes and others)?
- If yes, is such tweaking fine with you or would you prefer tarantool to adapt and throttle requests and/or force syncing WAL to disk in case WAL doesn't catch up?

**Comment by funny-falcon** [comment]
As far as I remember, there were no stalls on memtx.

**Comment by kyukhin** [comment]
Issue should be resolved in recent patches to 1.10.2. Could you pls give it a try?

**Comment by locker** [comment]
Please note that currently transactions are throttled only if there's a risk of long stalls caused by hitting `box.cfg.vinyl_memory` limit (#1862). For now, we don't take into account compaction progress, which means that the read and space amplification may grow if the write load is high. The only way to resolve it now is increase the number of writer threads (`box.cfg.vinyl_write_threads`), however we're planning to fix it soon (#3721).

**Comment by kyukhin** [comment]
Done.
